### PR TITLE
Invoices

### PR DESCRIPTION
**Pull request by ankrgyl**
This change introduces the token classification head as an option for `LayoutLMForQuestionAnswering`, which is leveraged by our new invoice models (but usable for any document type).

I did some refactoring, since we now reference a model that has a different model config type than `layoutlm`, and need to propagate word indices instead of just a start/end.

### CODE REVIEW

File: src/docquery/ext/config.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+""" LayoutLM model configuration"""
+from transformers import LayoutLMConfig
+
+
+class LayoutLMTokenClassifierConfig(LayoutLMConfig):
+    model_type = "layoutlm-tc"
+
+    def __init__(
+        self,
+        # New stuff added for DocQuery
```
Comment (rstebbing) [inline src/docquery/ext/config.py:24]: nit: Can you include a docstring from https://github.com/huggingface/transformers/blob/0a632f076d6b275690176b79c64c5559e1240b05/src/transformers/models/layoutlm/configuration_layoutlm.py#L40 and include explanations for `token_classification`, `token_classifier_reduction`, and `token_classifier_constant` there? @jeffsrobertson Are you able help with the explanations?

File: src/docquery/ext/pipeline.py [hunk 2]
```diff
@@ -417,17 +415,31 @@ def postprocess_extractive_qa(
...
+                answer_word_ids = list(unique_everseen([word_ids[i] for i in token_indices]))
+                if len(answer_word_ids) > 0 and answer_word_ids[0] is not None and answer_word_ids[-1] is not None:
                     answers.append(
                         {
                             "score": float(score),
-                            "answer": " ".join(words[word_start : word_end + 1]),
-                            "start": word_start,
-                            "end": word_end,
+                            "answer": " ".join(words[i] for i in answer_word_ids),
+                            "words": answer_word_ids,
```
Comment (rstebbing) [inline src/docquery/ext/pipeline.py:441]: nit: Perhaps this should be `word_ids` or even `answer_word_ids`? I think it's a bit surprising to have the answer `words` be word IDs that's all. (Notably, too, `word_ids` and `words` are the fields on each of the `model_outputs`.)

File: src/docquery/pipeline.py [hunk 3]
```diff
@@ -66,9 +76,10 @@ def get_pipeline(checkpoint=None, revision=None, device=None):
     else:
         model = checkpoint
 
-    pipeline_kwargs = {}
     if config.model_type == "vision-encoder-decoder":
         pipeline_kwargs["feature_extractor"] = model
+    elif config.model_type == "layoutlm-tc":
+        pipeline_kwargs["max_answer_len"] = 1000  # Let the token classifier filter things out
```
Comment (rstebbing) [inline src/docquery/pipeline.py:82]: nit: Is there any way to have this depend on an attribute from the `config` instead of it being hard-coded?

File: src/docquery/ext/model.py [hunk 4]
```diff
@@ -1,27 +1,34 @@
...
-import math
+from dataclasses import dataclass
 from typing import Optional, Tuple, Union
 
 import torch
 from torch import nn
-from torch.nn import CrossEntropyLoss
-from transformers.modeling_outputs import QuestionAnsweringModelOutput
-from transformers.models.layoutlm import LayoutLMModel, LayoutLMPreTrainedModel
+from torch.nn import CrossEntropyLoss, Linear
```
Comment (rstebbing) [inline src/docquery/ext/model.py:6]: nit: We use `nn.Linear` to within this file too. Perhaps we can just use `nn.CrossEntropyLoss` and `nn.Linear` and avoid the separate import?

File: src/docquery/ext/model.py [hunk 5]
```diff
@@ -133,14 +148,49 @@ def forward(
...
+
+
+# This is a thin wrapper around LayoutLMTokenClassifierForQuestionAnswering that simply instantiates
+# a default value for token_classification. Once we update transformers to be a version that
+# includes the upstremed LayoutLMForQuestionAnswering class, we can remove this.
+class LayoutLMForQuestionAnswering(LayoutLMTokenClassifierForQuestionAnswering):
+    config_class = LayoutLMConfig
+
+    def __init__(self, config, **kwargs):
+        config.token_classification = False
```
Comment (rstebbing) [inline src/docquery/ext/model.py:195]: nit: Use `config = copy.deepcopy(config)` (e.g. like in https://github.com/huggingface/transformers/blob/0a632f076d6b275690176b79c64c5559e1240b05/src/transformers/models/bart/modeling_bart.py#L1713).

File: src/docquery/ext/pipeline.py [hunk 6]
```diff
@@ -417,17 +415,31 @@ def postprocess_extractive_qa(
...
+                            # TODO: This is slightly inaccurate, since it is only the first page
```
Comment (rstebbing) [inline src/docquery/ext/pipeline.py:442]: nit: Why is this inaccurate? Aren't these words all on a single page as per:
```
for page_idx, (image, word_boxes) in enumerate(input["pages"])
```
?

**Review by rstebbing — COMMENTED** [review]
Nice :+1:! Thanks for putting this up!

File: src/docquery/ext/config.py (hunk 1 above)
Comment (ankrgyl) [reply src/docquery/ext/config.py:24]: Good idea. Added the basic parts and left TODOs where I need @jeffsrobertson's help.

File: src/docquery/ext/itertools.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+from itertools import filterfalse
+
+
+def unique_everseen(iterable, key=None):
+    "List unique elements, preserving order. Remember all elements ever seen."
```
[the remark this answers is not in the document]
Comment (ankrgyl) [reply src/docquery/ext/itertools.py:5]: Yep

File: src/docquery/ext/model.py (hunk 4 above)
Comment (ankrgyl) [reply src/docquery/ext/model.py:6]: Good idea, fixed.

File: src/docquery/ext/model.py (hunk 5 above)
Comment (ankrgyl) [reply src/docquery/ext/model.py:195]: Good catch

File: src/docquery/ext/pipeline.py (hunk 6 above)
Comment (ankrgyl) [reply src/docquery/ext/pipeline.py:442]: Good point, removed. I was thinking about a different implementation.

File: src/docquery/ext/pipeline.py (hunk 2 above)
Comment (ankrgyl) [reply src/docquery/ext/pipeline.py:441]: I'm a bit confused -- it is returning `answer_word_ids`. I think the best UX here would be to actually return the `[(word, box)]` (same format as the user input), so that it would work whether or not they ran OCR themselves. But I'd like to do that separately because it will complicate the tests a bit and I want to work out how the demo looks/feels before making another big change.

File: src/docquery/pipeline.py (hunk 3 above)
Comment (ankrgyl) [reply src/docquery/pipeline.py:82]: Sorry I forgot to reply to this. I don't think so, or at least it would be "un-kosher" since it's not a property of the model (but rather the post-processing/pipeline code). I think we could incorporate it inside the pipeline though, based on which model it is. The upstreamed version of the pipeline (in transformers) already does that for a few different types of models.

Comment (rstebbing) [reply src/docquery/pipeline.py:82]: Yeah, it would seem more sensible to accept `None` here and have that treated optionally in `decode_spans` (https://github.com/impira/docquery/blob/7a862d91be63ad56e6f6627395c3cdbcb423b405/src/docquery/ext/qa_helpers.py#L62). The kind of change I had in mind, FWIW, is something like:
```
# Disable truncating the answer in `decode_spans` by setting `max_answer_len` to a value
# greater than the maximum number of tokens.
pipeline_kwargs["max_answer_len"] = tokenizer.model_max_length + 1
```

Comment (ankrgyl) [reply src/docquery/pipeline.py:82]: Will update

File: src/docquery/ext/config.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" LayoutLM model configuration"""
+from transformers import LayoutLMConfig
+
+
+class LayoutLMTokenClassifierConfig(LayoutLMConfig):
```
Comment (jeffsrobertson) [inline src/docquery/ext/config.py:19]: Nice, I like that you baked the token classifier hyperparams directly into the config 👍

File: src/docquery/ext/model.py [hunk 9]
```diff
@@ -128,19 +143,55 @@ def forward(
...
+
+# This is a thin wrapper around LayoutLMTokenClassifierForQuestionAnswering that simply instantiates
+# a default value for token_classification. Once we update transformers to be a version that
+# includes the upstremed LayoutLMForQuestionAnswering class, we can remove this.
+class LayoutLMForQuestionAnswering(LayoutLMTokenClassifierForQuestionAnswering):
+    config_class = LayoutLMConfig
+
+    def __init__(self, config, **kwargs):
+        config = copy.deepcopy(config)
+        config.token_classification = False
```
Comment (jeffsrobertson) [inline src/docquery/ext/model.py:196]: Is this not covered by setting the default value for `token_classification` within `LayoutLMTokenClassifierConfig`?

**Review by jeffsrobertson — APPROVED** [review]
LGTM, nice change!

File: src/docquery/ext/model.py (hunk 9 above)
Comment (ankrgyl) [reply src/docquery/ext/model.py:196]: It is, this was just a way of being explicit (e.g. if we set it to true by default in the config class, which kind of makes sense given its purpose).

File: src/docquery/ext/config.py (hunk 8 above)
Comment (ankrgyl) [reply src/docquery/ext/config.py:19]: Cool!

File: src/docquery/ext/config.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-12):
+            The epsilon used by the layer normalization layers.
+        max_2d_position_embeddings (`int`, *optional*, defaults to 1024):
+            The maximum value that the 2D position embedding might ever used. Typically set this to something large
+            just in case (e.g., 1024).
+        token_classification (`bool, *optional*, defaults to False):
+            Whether to include an additional token classification head in question answering
+        token_classifier_reduction (`str`, *optional*, defaults to "mean")
+            # TODO
```
Comment (jeffsrobertson) [inline src/docquery/ext/config.py:61]: `Specifies the reduction to apply to the output of the cross entropy loss for the token classifier head during training. Options are: 'none' | 'mean' | 'sum'. 'none': no reduction will be applied, 'mean': the weighted mean of the output is taken, 'sum': the output will be summed.`

File: src/docquery/ext/config.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        token_classifier_constant (`float`, *optional*, defaults to 1.0)
+            # TODO
```
Comment (jeffsrobertson) [inline src/docquery/ext/config.py:63]: `Coefficient for the token classifier head's contribution to the total loss. A larger value means that the model will prioritize learning the token classifier head during training.`

Comment (ankrgyl) [reply src/docquery/ext/config.py:63]: THank you!

File: src/docquery/ext/config.py (hunk 10 above)
Comment (ankrgyl) [reply src/docquery/ext/config.py:61]: Thank you!
